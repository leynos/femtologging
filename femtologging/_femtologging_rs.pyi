from typing import Any as _Any, Literal, Union

FemtoLevel: _Any
LevelName = Literal["TRACE", "DEBUG", "INFO", "WARN", "WARNING", "ERROR", "CRITICAL"]
LevelArg = Union[LevelName, FemtoLevel]

FemtoLogger: _Any
FemtoHandler: _Any
FemtoStreamHandler: _Any
FemtoFileHandler: _Any
FemtoRotatingFileHandler: _Any
StreamHandlerBuilder: _Any
FileHandlerBuilder: _Any
RotatingFileHandlerBuilder: _Any
ConfigBuilder: _Any
LoggerConfigBuilder: _Any
FormatterBuilder: _Any
LevelFilterBuilder: _Any
NameFilterBuilder: _Any
FilterBuildError: type[Exception]
HandlerConfigError: type[Exception]
HandlerIOError: type[Exception]

hello: _Any
get_logger: _Any
reset_manager_py: _Any
