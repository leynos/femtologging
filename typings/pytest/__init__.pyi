from typing import (
    Any,
    Callable,
    ContextManager,
    Pattern,
    Type,
    TypeVar,
    overload,
    ParamSpec,
)

P = ParamSpec("P")
R = TypeVar("R", bound=BaseException)

@overload
def fixture(func: Callable[P, R]) -> Callable[P, R]: ...
@overload
def fixture(
    *,
    scope: str | None = ...,
    autouse: bool | None = ...,
    params: list[Any] | None = ...,
    ids: list[str] | Callable[[Any], str] | None = ...,
    name: str | None = ...,
) -> Callable[[Callable[P, R]], Callable[P, R]]: ...
def raises(
    exc: Type[R],
    match: str | Pattern[str] | None = ...,
    *,
    msg: str | None = ...,
) -> ContextManager[R]: ...

class mark:
    @staticmethod
    def parametrize(
        argnames: str | list[str],
        argvalues: list[Any] | tuple[Any, ...] | list[tuple[Any, ...]],
        *,
        ids: list[str] | Callable[[Any], str] | None = ...,
        indirect: bool | list[str] = ...,
        **kwargs: Any,
    ) -> Callable[[Callable[..., Any]], Callable[..., Any]]: ...
