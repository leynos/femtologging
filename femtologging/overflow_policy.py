from __future__ import annotations

from enum import Enum


class OverflowPolicy(Enum):
    """Behaviour when a handler queue is full."""

    DROP = "drop"
    BLOCK = "block"
    TIMEOUT = "timeout"

    def __str__(self) -> str:
        return self.value
