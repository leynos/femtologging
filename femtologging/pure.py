from __future__ import annotations


class FemtoLogger:
    """Simplistic Python implementation used when the Rust module is missing."""

    def __init__(self, name: str) -> None:
        self.name = name

    def log(self, level: str, message: str) -> str:
        """Return the formatted log message."""
        return f"{self.name} [{level}] {message}"


def hello() -> str:
    """Return a friendly greeting from Python."""
    return "hello from Python"


__all__ = ["FemtoLogger", "hello"]
