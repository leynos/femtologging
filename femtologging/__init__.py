"""femtologging package."""

from __future__ import annotations

# Import the Rust extension packaged under this module's namespace first
# to keep imports at the top for linters.
from . import _femtologging_rs as rust
from .overflow_policy import OverflowPolicy
from .config import dictConfig
import logging
import sys
from dataclasses import dataclass
from typing import (
    Callable,
    Iterable,
    TextIO,
    cast,
    overload,
)


hello = rust.hello
FemtoLogger = rust.FemtoLogger
get_logger = rust.get_logger
reset_manager = rust.reset_manager_py
FemtoHandler = rust.FemtoHandler
FemtoStreamHandler = rust.FemtoStreamHandler
FemtoFileHandler = rust.FemtoFileHandler
FemtoRotatingFileHandler = rust.FemtoRotatingFileHandler
HandlerOptions = rust.HandlerOptions
ROTATION_VALIDATION_MSG = rust.ROTATION_VALIDATION_MSG
StreamHandlerBuilder = rust.StreamHandlerBuilder
FileHandlerBuilder = rust.FileHandlerBuilder
RotatingFileHandlerBuilder = rust.RotatingFileHandlerBuilder
ConfigBuilder = rust.ConfigBuilder
LoggerConfigBuilder = rust.LoggerConfigBuilder
FormatterBuilder = rust.FormatterBuilder
LevelFilterBuilder = rust.LevelFilterBuilder
NameFilterBuilder = rust.NameFilterBuilder
FilterBuildError = rust.FilterBuildError
HandlerConfigError = rust.HandlerConfigError
HandlerIOError = rust.HandlerIOError
_force_rotating_fresh_failure = getattr(
    rust, "force_rotating_fresh_failure_for_test", None
)
_clear_rotating_fresh_failure = getattr(
    rust, "clear_rotating_fresh_failure_for_test", None
)

if callable(_force_rotating_fresh_failure) and callable(_clear_rotating_fresh_failure):
    _force_rotating_fresh_failure_for_test = cast(
        Callable[[int, str | None], None],
        _force_rotating_fresh_failure,
    )
    _clear_rotating_fresh_failure_for_test = cast(
        Callable[[], None],
        _clear_rotating_fresh_failure,
    )
else:
    # Feature disabled: expose no-ops that fail loudly when invoked.

    def _force_rotating_fresh_failure_for_test(
        count: int, reason: str | None = None
    ) -> None:
        raise RuntimeError(
            "rotating fresh-failure hook requires the extension built with the 'python' feature"
        )

    def _clear_rotating_fresh_failure_for_test() -> None:
        return


@dataclass
class BasicConfig:
    """Configuration parameters for basicConfig()."""

    level: str | int | None = None
    filename: str | None = None
    stream: TextIO | None = None
    force: bool = False
    handlers: Iterable[FemtoHandler] | None = None


@overload
def basicConfig(config: BasicConfig, /) -> None: ...


@overload
def basicConfig(**kwargs: object) -> None: ...


def basicConfig(config: BasicConfig | None = None, /, **kwargs: object) -> None:
    """Configure the root logger using the builder API.

        Parameters mirror ``logging.basicConfig`` but currently only a subset is
        supported. ``config`` may be a :class:`BasicConfig` instance; if provided,
        its values take precedence over individual parameters. ``level`` may be a
        string or numeric value understood by the standard :mod:`logging` module.
        ``filename`` configures a :class:`FemtoFileHandler`; otherwise a
        :class:`FemtoStreamHandler` targeting ``stderr`` is installed. ``stream``
        may be ``sys.stdout`` to redirect output. ``force`` removes any existing
        handlers from the root logger before applying the new configuration.
        ``handlers`` allows attaching pre‑constructed handlers directly.

        Parameters
        ----------
        config : BasicConfig, optional
            Configuration dataclass providing parameters for ``basicConfig``.

        Other Parameters
        ----------------
        level : str or int, optional
            Logging level. Accepts case-insensitive "TRACE", "DEBUG", "INFO", "WARN",
    "WARNING", "ERROR", and "CRITICAL". "WARN" and "WARNING" are equivalent.
        filename : str, optional
            File to write logs to.
        stream : TextIO, optional
            ``sys.stdout`` or ``sys.stderr``.
        force : bool, default False
            Remove any existing handlers before configuring.
        handlers : Iterable[FemtoHandler], optional
            Pre‑constructed handlers to attach.

        Examples
        --------
        Using a dataclass::

            cfg = BasicConfig(level="INFO")
            basicConfig(cfg)

        Using individual parameters::

            basicConfig(level="INFO")

        Notes
        -----
        ``format`` and ``datefmt`` are intentionally unsupported until formatter
        customisation is implemented.
    """
    allowed = {"level", "filename", "stream", "force", "handlers"}
    unknown = set(kwargs) - allowed
    if unknown:
        name = next(iter(unknown))
        raise TypeError(f"basicConfig() got an unexpected keyword argument {name!r}")

    level: str | int | None
    filename: str | None
    stream: TextIO | None
    force: bool
    handlers: Iterable[FemtoHandler] | None
    if config is not None:
        level = (
            config.level
            if config.level is not None
            else cast(str | int | None, kwargs.get("level"))
        )
        filename = (
            config.filename
            if config.filename is not None
            else cast(str | None, kwargs.get("filename"))
        )
        stream = (
            config.stream
            if config.stream is not None
            else cast(TextIO | None, kwargs.get("stream"))
        )
        force = bool(config.force)
        handlers = (
            config.handlers
            if config.handlers is not None
            else cast(Iterable[FemtoHandler] | None, kwargs.get("handlers"))
        )
    else:
        level = cast(str | int | None, kwargs.get("level"))
        filename = cast(str | None, kwargs.get("filename"))
        stream = cast(TextIO | None, kwargs.get("stream"))
        force = bool(kwargs.get("force", False))
        handlers = cast(Iterable[FemtoHandler] | None, kwargs.get("handlers"))

    _validate_basic_config_params(filename, stream, handlers)

    root = get_logger("root")
    if force:
        root.clear_handlers()

    _configure_handlers(root, handlers, filename, stream)

    _set_logger_level(root, level)


def _has_conflicting_handler_params(
    handlers: Iterable[FemtoHandler] | None,
    filename: str | None,
    stream: TextIO | None,
) -> bool:
    """Check if handlers conflict with filename or stream parameters."""
    return handlers is not None and (filename is not None or stream is not None)


def _validate_basic_config_params(
    filename: str | None,
    stream: TextIO | None,
    handlers: Iterable[FemtoHandler] | None,
) -> None:
    """Validate ``basicConfig`` parameters."""
    if filename and stream:
        raise ValueError("Cannot specify both `filename` and `stream`")

    if _has_conflicting_handler_params(handlers, filename, stream):
        msg = "Cannot specify `handlers` with `filename` or `stream`"
        raise ValueError(msg)

    if stream not in (None, sys.stdout, sys.stderr):
        raise ValueError(
            f"stream must be sys.stdout or sys.stderr, got {type(stream)!r}: {stream!r}"
        )


def _configure_handlers(
    root: FemtoLogger,
    handlers: Iterable[FemtoHandler] | None,
    filename: str | None,
    stream: TextIO | None,
) -> None:
    """Attach or build handlers for the root logger."""
    if handlers is not None:
        for h in handlers:
            root.add_handler(h)
    else:
        _build_and_configure_handler(filename, stream)


def _build_and_configure_handler(
    filename: str | None,
    stream: TextIO | None,
) -> None:
    """Build a handler via the builder API and install it."""
    builder = ConfigBuilder()

    handler_id = "basic_config_handler"
    handler = _create_handler_builder(filename, stream)
    builder.with_handler(handler_id, handler)

    logger_cfg = LoggerConfigBuilder().with_handlers([handler_id])
    builder.with_root_logger(logger_cfg)

    builder.build_and_init()


def _create_handler_builder(
    filename: str | None,
    stream: TextIO | None,
) -> FileHandlerBuilder | StreamHandlerBuilder:
    """Create a handler builder for ``basicConfig``."""
    if filename:
        return FileHandlerBuilder(filename)
    if stream is sys.stdout:
        return StreamHandlerBuilder.stdout()
    return StreamHandlerBuilder.stderr()


def _set_logger_level(root: FemtoLogger, level: str | int | None) -> None:
    """Set the root logger level if provided."""
    if level is not None:
        lvl = logging.getLevelName(level) if isinstance(level, int) else level
        root.set_level(lvl)


__all__ = [
    "FemtoHandler",
    "FemtoLogger",
    "get_logger",
    "reset_manager",
    "FemtoStreamHandler",
    "FemtoFileHandler",
    "FemtoRotatingFileHandler",
    "HandlerOptions",
    "ROTATION_VALIDATION_MSG",
    "StreamHandlerBuilder",
    "FileHandlerBuilder",
    "RotatingFileHandlerBuilder",
    "LevelFilterBuilder",
    "NameFilterBuilder",
    "FilterBuildError",
    "ConfigBuilder",
    "LoggerConfigBuilder",
    "FormatterBuilder",
    "HandlerConfigError",
    "HandlerIOError",
    "OverflowPolicy",
    "BasicConfig",
    "basicConfig",
    "dictConfig",
    "hello",
]
