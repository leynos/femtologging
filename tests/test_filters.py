from __future__ import annotations

import collections.abc as cabc
import pytest
from pytest_bdd import given, parsers, scenarios, then, when
from syrupy import SnapshotAssertion

from femtologging import (
    ConfigBuilder,
    LoggerConfigBuilder,
    StreamHandlerBuilder,
    LevelFilterBuilder,
    NameFilterBuilder,
    get_logger,
    reset_manager,
    FilterBuildError,
)


@pytest.fixture(autouse=True)
def reset_logger_state() -> cabc.Iterator[None]:
    reset_manager()
    yield
    reset_manager()


@given("a ConfigBuilder", target_fixture="config_builder")
def config_builder() -> ConfigBuilder:
    return ConfigBuilder()


@when(parsers.parse('I add stream handler "{hid}" targeting "{target}"'))
def add_stream_handler(config_builder: ConfigBuilder, hid: str, target: str) -> None:
    handler = (
        StreamHandlerBuilder.stderr()
        if target.lower() == "stderr"
        else StreamHandlerBuilder.stdout()
    )
    config_builder.with_handler(hid, handler)


@when(parsers.parse('I add level filter "{fid}" with max level "{level}"'))
def add_level_filter(config_builder: ConfigBuilder, fid: str, level: str) -> None:
    filt = LevelFilterBuilder().with_max_level(level)
    config_builder.with_filter(fid, filt)


@when(parsers.parse('I add name filter "{fid}" with prefix "{prefix}"'))
def add_name_filter(config_builder: ConfigBuilder, fid: str, prefix: str) -> None:
    filt = NameFilterBuilder().with_prefix(prefix)
    config_builder.with_filter(fid, filt)


@when(
    parsers.parse(
        'I add logger "{name}" with handler "{handler}" and filter "{filter_id}"'
    )
)
def add_logger_with_filter(
    config_builder: ConfigBuilder, name: str, handler: str, filter_id: str
) -> None:
    logger = LoggerConfigBuilder().with_handlers([handler]).with_filters([filter_id])
    config_builder.with_logger(name, logger)


@when(parsers.parse('I add logger "{name}" with filter "{filter_id}"'))
def add_logger_only_filter(
    config_builder: ConfigBuilder, name: str, filter_id: str
) -> None:
    logger = LoggerConfigBuilder().with_filters([filter_id])
    config_builder.with_logger(name, logger)


@when(parsers.parse('I set root logger with level "{level}"'))
def set_root(config_builder: ConfigBuilder, level: str) -> None:
    root = LoggerConfigBuilder().with_level(level)
    config_builder.with_root_logger(root)


@then("the configuration matches snapshot")
def configuration_matches_snapshot(
    config_builder: ConfigBuilder, snapshot: SnapshotAssertion
) -> None:
    assert config_builder.as_dict() == snapshot
    config_builder.build_and_init()


@then(parsers.parse('logger "{name}" emits "{level}"'))
def logger_emits(name: str, level: str) -> None:
    logger = get_logger(name)
    assert logger.log(level, "msg") is not None


@then(parsers.parse('logger "{name}" suppresses "{level}"'))
def logger_suppresses(name: str, level: str) -> None:
    logger = get_logger(name)
    assert logger.log(level, "msg") is None


@then("building the configuration fails")
def build_fails(config_builder: ConfigBuilder) -> None:
    with pytest.raises(FilterBuildError):
        config_builder.build_and_init()


@then(parsers.parse('building the configuration fails with error containing "{msg}"'))
def build_fails_with_message(config_builder: ConfigBuilder, msg: str) -> None:
    with pytest.raises(FilterBuildError) as excinfo:
        config_builder.build_and_init()
    assert msg in str(excinfo.value)


@then(
    parsers.parse(
        'building the configuration fails with value error containing "{msg}"'
    )
)
def build_fails_with_value_error(config_builder: ConfigBuilder, msg: str) -> None:
    with pytest.raises(ValueError) as excinfo:
        config_builder.build_and_init()
    assert msg in str(excinfo.value)


scenarios("features/filters.feature")


def test_logger_with_multiple_filters() -> None:
    cb = (
        ConfigBuilder()
        .with_filter("lvl", LevelFilterBuilder().with_max_level("INFO"))
        .with_filter("name", NameFilterBuilder().with_prefix("multi"))
        .with_logger(
            "multi",
            LoggerConfigBuilder().with_filters(["lvl", "name"]),
        )
        .with_root_logger(LoggerConfigBuilder().with_level("INFO"))
    )
    cb.build_and_init()
    logger = get_logger("multi")
    assert logger.log("INFO", "emit") is not None
    assert logger.log("DEBUG", "suppress") is None


@pytest.mark.parametrize(
    ("_scenario", "first_filter", "second_filter"),
    [
        (
            "remove_all_filters",
            ("lvl", LevelFilterBuilder().with_max_level("DEBUG")),
            None,
        ),
        (
            "replace_with_name_filter",
            ("lvl", LevelFilterBuilder().with_max_level("DEBUG")),
            ("name", NameFilterBuilder().with_prefix("core")),
        ),
    ],
)
def test_reconfig_replaces_filters(
    _scenario: str,
    first_filter: tuple[str, LevelFilterBuilder],
    second_filter: tuple[str, NameFilterBuilder] | None,
) -> None:
    cb = (
        ConfigBuilder()
        .with_filter(first_filter[0], first_filter[1])
        .with_logger("core", LoggerConfigBuilder().with_filters([first_filter[0]]))
        .with_root_logger(LoggerConfigBuilder().with_level("DEBUG"))
    )
    cb.build_and_init()
    logger = get_logger("core")
    assert logger.log("ERROR", "drop") is None

    reconfig = ConfigBuilder().with_root_logger(
        LoggerConfigBuilder().with_level("DEBUG")
    )
    if second_filter is not None:
        reconfig = reconfig.with_filter(second_filter[0], second_filter[1]).with_logger(
            "core", LoggerConfigBuilder().with_filters([second_filter[0]])
        )
    else:
        reconfig = reconfig.with_logger("core", LoggerConfigBuilder())
    reconfig.build_and_init()

    logger_after = get_logger("core")
    assert logger_after.log("ERROR", "emit") is not None


def test_reconfig_with_unknown_filter_preserves_previous_filters() -> None:
    cb = (
        ConfigBuilder()
        .with_filter("lvl", LevelFilterBuilder().with_max_level("DEBUG"))
        .with_logger("core", LoggerConfigBuilder().with_filters(["lvl"]))
        .with_root_logger(LoggerConfigBuilder().with_level("DEBUG"))
    )
    cb.build_and_init()
    logger = get_logger("core")
    assert logger.log("ERROR", "drop") is None

    bad = (
        ConfigBuilder()
        .with_logger("core", LoggerConfigBuilder().with_filters(["missing"]))
        .with_root_logger(LoggerConfigBuilder().with_level("DEBUG"))
    )
    with pytest.raises(ValueError):
        bad.build_and_init()

    logger_after = get_logger("core")
    assert logger_after.log("ERROR", "still drop") is None


def test_filter_clearing() -> None:
    cb = (
        ConfigBuilder()
        .with_filter("lvl", LevelFilterBuilder().with_max_level("DEBUG"))
        .with_logger("core", LoggerConfigBuilder().with_filters(["lvl"]))
        .with_root_logger(LoggerConfigBuilder().with_level("INFO"))
    )
    cb.build_and_init()
    logger = get_logger("core")
    assert logger.log("INFO", "drop") is None
    logger.clear_filters()
    assert logger.log("INFO", "emit") is not None


def test_multiple_filters_clearing() -> None:
    cb = (
        ConfigBuilder()
        .with_filter("lvl", LevelFilterBuilder().with_max_level("DEBUG"))
        .with_filter("name", NameFilterBuilder().with_prefix("other"))
        .with_logger(
            "core",
            LoggerConfigBuilder().with_level("DEBUG").with_filters(["lvl", "name"]),
        )
        .with_root_logger(LoggerConfigBuilder().with_level("DEBUG"))
    )
    cb.build_and_init()
    logger = get_logger("core")
    assert logger.log("ERROR", "blocked by level filter") is None
    assert logger.log("DEBUG", "blocked by name") is None
    logger.clear_filters()
    assert logger.log("ERROR", "allowed now") is not None
    assert logger.log("DEBUG", "also allowed") is not None
