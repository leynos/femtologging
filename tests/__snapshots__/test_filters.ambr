# serializer version: 1
# name: test_level_filter_suppresses_high_severity
  dict({
    'disable_existing_loggers': False,
    'filters': dict({
      'only_info': dict({
        'max_level': 'INFO',
      }),
    }),
    'handlers': dict({
      'console': dict({
        'target': 'stderr',
      }),
    }),
    'loggers': dict({
      'core': dict({
        'filters': list([
          'only_info',
        ]),
        'handlers': list([
          'console',
        ]),
      }),
    }),
    'root': dict({
      'level': 'DEBUG',
    }),
    'version': 1,
  })
# ---
# name: test_name_filter_matches_prefix
  dict({
    'disable_existing_loggers': False,
    'filters': dict({
      'core_only': dict({
        'prefix': 'core',
      }),
    }),
    'handlers': dict({
      'console': dict({
        'target': 'stderr',
      }),
    }),
    'loggers': dict({
      'core.child': dict({
        'filters': list([
          'core_only',
        ]),
        'handlers': list([
          'console',
        ]),
      }),
      'other': dict({
        'filters': list([
          'core_only',
        ]),
        'handlers': list([
          'console',
        ]),
      }),
    }),
    'root': dict({
      'level': 'DEBUG',
    }),
    'version': 1,
  })
# ---
